Disable the (name+num)+num = name+newnum optimization.

sunfishcode opened the pull request:
This optimization reassociates addition and can introduce situations where intermediate computation wraps around the address space, creating inefficiency in JS engines which are optimized for no-wraparound address arithmetic.

In its place, the fastcomp backend is being enhanced in https://github.com/kripken/emscripten-fastcomp/pull/98 to fold getelementptrs with constant bases in more cases, which is the main case where the above optimization happens in practice.

kripken commented:
This should require updates to `other.test_js_optimizer` (if it doesn't, we need to add tests that do break on a change like this!)

I would run at least `random200` and `other` to get good coverage for this.

juj commented:
For reference, this is the inefficiency in JS engines that was mentioned in OP: https://bugzilla.mozilla.org/show_bug.cgi?id=1159404

sunfishcode commented:
I updated the patch to modify the js optimizer too, and that seemed to find an existing test that failed. Is the native optimizer not getting tested, or did I do something wrong? Anyway, I updated the test now.

kripken commented:
Some parts were not fully tested, I am adding patches now. I see that your work here fixes the tests, so things like fine.

However I seem to see benchmark regressions, e.g. on zlib. Do you see those?
